File: iosApp/iosApp/reown/mappers.swift
import App
import ReownWalletKit

func mapProposal(
    proposal: ReownWalletKit.Session.Proposal,
    context: ReownWalletKit.VerifyContext?
) -> App.SessionProposal {
    return App.SessionProposal(
        id: Int64(proposal.id.hashValue),
        pairingTopic: proposal.pairingTopic,
        name: proposal.proposer.name,
        url: proposal.proposer.url,
        icon: proposal.proposer.icons.first,
        description: proposal.proposer.description,
        verifyContext: mapContext(context: context),
        required: mapProposalNamespace(namespace: proposal.requiredNamespaces[ReownAdapterKt.EIP155]),
        optional: mapProposalNamespace(namespace: proposal.optionalNamespaces?[ReownAdapterKt.EIP155])
    )
}

func mapAuthentication(
    authentication: ReownWalletKit.AuthenticationRequest,
    context: ReownWalletKit.VerifyContext?
) -> App.AuthenticationRequest {
    return App.AuthenticationRequest(
        id: authentication.id.integer,
        pairingTopic: authentication.topic,
        name: authentication.requester.name,
        url: authentication.requester.url,
        icon: authentication.requester.icons.first,
        description: authentication.requester.description,
        verifyContext: mapContext(context: context)
    )
}

func mapRequest(
    request: ReownWalletKit.Request,
    session: ReownWalletKit.Session,
    context: ReownWalletKit.VerifyContext?
) -> App.SessionRequest {
    return App.SessionRequest(
        id: request.id.integer,
        sessionTopic: request.topic,
        name: session.peer.name,
        url: session.peer.url,
        icon: session.peer.icons.first,
        description: session.peer.description,
        chainId: KotlinULong(value: UInt64(request.chainId.reference)!),
        method: request.method,
        params: request.params.stringRepresentation,
        verifyContext: mapContext(context: context)
    )
}

func mapSession(
    session: ReownWalletKit.Session
) -> App.Session {
    return App.Session(
        sessionTopic: session.topic,
        name: session.peer.name,
        url: session.peer.url,
        icon: session.peer.icons.first,
        description: session.peer.description,
        required: mapProposalNamespace(namespace: session.requiredNamespaces[ReownAdapterKt.EIP155]),
        optional: mapProposalNamespace(namespace: nil),
        approved: mapSessionNamespace(namespace: session.namespaces[ReownAdapterKt.EIP155])
    )
}

func mapProposalNamespace(
    namespace : ProposalNamespace?
) -> App.SessionProposal.Namespace {
    return App.SessionProposal.Namespace(
        chains: namespace?.chains?.map({ch in ch.absoluteString}) ?? [],
        methods: Array(namespace?.methods ?? []),
        events: Array(namespace?.events ?? [])
    )
}

func mapSessionNamespace(
    namespace : SessionNamespace?
) -> App.Session.Namespace {
    return App.Session.Namespace(
        chains: namespace?.chains?.map({ch in ch.absoluteString}) ?? [],
        accounts: namespace?.accounts.map({acc in acc.absoluteString}) ?? [],
        methods: Array(namespace?.methods ?? []),
        events: Array(namespace?.events ?? [])
    )
}

func mapContext(
    context: ReownWalletKit.VerifyContext?
) -> App.VerifyContext {
    return switch context?.validation {
    case .some(ReownWalletKit.VerifyContext.ValidationStatus.invalid):
        App.VerifyContext.mismatch
    case .some(ReownWalletKit.VerifyContext.ValidationStatus.scam):
        App.VerifyContext.threat
    case .some(ReownWalletKit.VerifyContext.ValidationStatus.unknown):
        App.VerifyContext.unverified
    case .some(ReownWalletKit.VerifyContext.ValidationStatus.valid):
        App.VerifyContext.match
    case .none:
        App.VerifyContext.unverified
    }
}

File: libs/reown/src/androidMain/kotlin/kosh/libs/reown/mappers.kt
package kosh.libs.reown

import com.reown.android.internal.common.signing.cacao.CacaoType
import com.reown.sign.client.Sign
import com.reown.walletkit.client.Wallet


internal fun mapProposal(
    proposal: Wallet.Model.SessionProposal,
    context: Wallet.Model.VerifyContext?,
) = SessionProposal(
    id = context?.id ?: proposal.proposerPublicKey.hashCode().toLong(),
    pairingTopic = proposal.pairingTopic,
    name = proposal.name,
    description = proposal.description,
    url = proposal.url,
    icon = proposal.icons.firstOrNull()?.toString(),
    verifyContext = mapContext(context),
    required = mapNamespace(proposal.requiredNamespaces[EIP155]),
    optional = mapNamespace(proposal.optionalNamespaces[EIP155]),
)


internal fun mapAuth(
    authenticate: Wallet.Model.SessionAuthenticate,
    context: Wallet.Model.VerifyContext?,
): AuthenticationRequest = AuthenticationRequest(
    id = authenticate.id,
    name = authenticate.participant.metadata?.name,
    description = authenticate.participant.metadata?.description,
    url = authenticate.participant.metadata?.url,
    icon = authenticate.participant.metadata?.icons?.firstOrNull()?.toString(),
    pairingTopic = authenticate.pairingTopic,
    verifyContext = mapContext(context)
)

internal fun mapRequest(
    request: Wallet.Model.SessionRequest,
    context: Wallet.Model.VerifyContext?,
): SessionRequest = SessionRequest(
    id = request.request.id,
    verifyContext = mapContext(context),
    sessionTopic = request.topic,
    name = request.peerMetaData?.name,
    description = request.peerMetaData?.description,
    url = request.peerMetaData?.url,
    icon = request.peerMetaData?.icons?.firstOrNull()?.toString(),
    chainId = request.chainId?.toULongOrNull(),
    method = request.request.method,
    params = request.request.params,
)

internal fun mapSession(
    session: Wallet.Model.Session,
): Session = Session(
    sessionTopic = session.topic,
    name = session.metaData?.name,
    description = session.metaData?.description,
    url = session.metaData?.url,
    icon = session.metaData?.icons?.firstOrNull()?.toString(),
    required = mapNamespace(session.requiredNamespaces[EIP155]),
    optional = mapNamespace((session.optionalNamespaces ?: mapOf())[EIP155]),
    approved = mapNamespace(session.namespaces[EIP155]),
)

internal fun mapNamespace(
    approvedNamespace: Wallet.Model.Namespace.Session?,
) = Session.Namespace(
    chains = approvedNamespace?.chains ?: listOf(),
    accounts = approvedNamespace?.accounts ?: listOf(),
    methods = approvedNamespace?.methods ?: listOf(),
    events = approvedNamespace?.events ?: listOf()
)

internal fun mapNamespace(
    approvedNamespace: Wallet.Model.Namespace.Proposal?,
) = SessionProposal.Namespace(
    chains = approvedNamespace?.chains ?: listOf(),
    methods = approvedNamespace?.methods ?: listOf(),
    events = approvedNamespace?.events ?: listOf()
)

internal fun mapContext(
    context: Wallet.Model.VerifyContext?,
): VerifyContext {
    if (context == null) return VerifyContext.Unverified
    return when (context.isScam) {
        true -> VerifyContext.Threat
        else -> when (context.validation) {
            Wallet.Model.Validation.VALID -> VerifyContext.Match
            Wallet.Model.Validation.INVALID -> VerifyContext.Mismatch
            Wallet.Model.Validation.UNKNOWN -> VerifyContext.Unverified
        }
    }
}

internal fun Sign.Model.SessionAuthenticate.toWallet(): Wallet.Model.SessionAuthenticate =
    Wallet.Model.SessionAuthenticate(id, topic, participant.toWallet(), payloadParams.toWallet())

internal fun Sign.Model.SessionAuthenticate.Participant.toWallet(): Wallet.Model.SessionAuthenticate.Participant =
    Wallet.Model.SessionAuthenticate.Participant(publicKey, metadata)

internal fun Sign.Model.PayloadParams.toWallet(): Wallet.Model.PayloadAuthRequestParams =
    Wallet.Model.PayloadAuthRequestParams(
        chains = chains,
        type = type ?: CacaoType.CAIP222.header,
        domain = domain,
        aud = aud,
        nonce = nonce,
        nbf = nbf,
        exp = exp,
        statement = statement,
        requestId = requestId,
        resources = resources,
        iat = iat
    )
